### PR TITLE
apiextensions-apiserver: add pkg/cmd/server/testing pkg for integration tests

### PR DESCRIPTION
**Pull request by sttts**
In analogy to kube-apiserver, this implements a uniform apiextensions-apiserver for integration tests which can be started using customized flags.

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+	if err := s.Complete(); err != nil {
+		return result, fmt.Errorf("failed to set default options: %v", err)
+	}
+	if err := s.Validate(); err != nil {
+		return result, fmt.Errorf("failed to validate options: %v", err)
+	}
+
+	t.Logf("runtime-config=%v", s.APIEnablement.RuntimeConfig)
+	t.Logf("Starting kube-apiserver on port %d...", s.RecommendedOptions.SecureServing.BindPort)
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go]: s/kube-apiserver/apiextensions-apiserver

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go]: 2018

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+limitations under the License.
+*/
+
+package server
+
+import (
+	"io"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apiextensions-apiserver/pkg/cmd/server/options"
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go]: nit: newline before the k8s.io import.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"time"
+
+	"github.com/spf13/pflag"
+	genericapiserver "k8s.io/apiserver/pkg/server"
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go]: nit: import order

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go]: done

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go]: done

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go]: done

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go]: done

**Comment by sttts** [comment]
@nikhita addressed your comments. ptal.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/server.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/server.go]: 2018

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go [hunk 6]
```diff
@@ -127,16 +106,3 @@ func NewCRDRESTOptionsGetter(etcdOptions genericoptions.EtcdOptions) genericregi
 
 	return ret
 }
-
-func (o CustomResourceDefinitionsServerOptions) RunCustomResourceDefinitionsServer(stopCh <-chan struct{}) error {
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go]: just curious, why remove this from here and add in `server.go`?

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go [hunk 7]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package testserver
+package fixtures
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go]: I think this file needs a 2017 -> 2018 bump too.

**Comment by nikhita** [comment]
can `test/integration/helpers.go` also go into fixtures?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go (hunk 6 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go]: RunCustomResourceDefinitionsServer exists in options.go is odd.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"io"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/apiextensions-apiserver/pkg/cmd/server/options"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+)
+
+func NewCommandStartCustomResourceDefinitionsServer(out, errOut io.Writer, stopCh <-chan struct{}) *cobra.Command {
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go]: Can we rename `NewCommandStartCustomResourceDefinitionsServer` to a better name? like `NewCustomResourceDefinitionsServerCommand` or `NewAPIExtensionsServerCommand`

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go]: NewServerCommand? The package gives the context.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go (hunk 7 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go]: It's just renamed. We don't change the year usually for that.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/server.go (hunk 5 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/server.go]: This is a rename that git does not detect.

**Comment by sttts** [comment]
> can test/integration/helpers.go also go into fixtures?

We moved those funcs out of the fixture directory because they were helper. Fixtures are usualy some partly constant test objects. I don't think the helpers really fit there.

**Comment by sttts** [comment]
@hzxuzhonghu @nikhita addressed your comments. ptal.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go (hunk 8 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/server.go]: sgtm

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+		return result, fmt.Errorf("failed to create config from options: %v", err)
+	}
+	server, err := config.Complete().New(genericapiserver.NewEmptyDelegate())
+	if err != nil {
+		return result, fmt.Errorf("failed to create server: %v", err)
+	}
+
+	go func(stopCh <-chan struct{}) {
+		if err := server.GenericAPIServer.PrepareRun().Run(stopCh); err != nil {
+			t.Errorf("kube-apiserver failed run: %v", err)
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go]: not kube-apiserver

**Comment by hzxuzhonghu** [comment]
Overall, it lgtm

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go (hunk 9 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go]: fixed

**Comment by nikhita** [comment]
:+1:

**Pull request merged**